File: 200-number-of-islands/number-of-islands.cpp
class Solution {
public:
    int n,m;
    
    void dfs(int row,int col,vector<vector<char>>&grid,vector<vector<int>>&visited)
    {
        visited[row][col] = 1;

        int delrow[] = {1,-1,0, 0};
        int delcol[] = {0 , 0 ,-1,1};

        for(int i = 0;i<4;i++)
        {
            int nrow = row + delrow[i];
            int ncol = col + delcol[i];

            if(nrow>=0 and nrow<n and ncol >=0 and
             ncol < m and grid[nrow][ncol] == '1' and !visited[nrow][ncol])
            {
               dfs(nrow,ncol,grid,visited);
            }
        }
    }
    int numIslands(vector<vector<char>>& grid) {
        n = grid.size();
        m = grid[0].size();
        vector<vector<int>>visited(n,vector<int>(m,0));
        int count = 0;
        for(int i = 0;i<n;i++)
        {
          for(int j = 0;j<m;j++)
          {
            if(grid[i][j] == '1' and !visited[i][j])
            {
                dfs(i,j,grid,visited);
                count++;
            }
          }
        }

        return count;
    }
};

File: 200-number-of-islands/number-of-islands.java
class Pair {
    int first;
    int second;

    public Pair(int first, int second) {
        this.first = first;
        this.second = second;
    }
}

class Solution {

    private void bfs(int i, int j, boolean visited[][], char[][] grid) {
        visited[i][j] = true;

        Queue<Pair> q = new LinkedList<Pair>();
        q.add(new Pair(i, j));

        int delRow[] = { 1, 0, -1, 0 };
        int delCol[] = { 0, -1, 0, 1 };

        while (!q.isEmpty()) {
            int row = q.peek().first;
            int col = q.peek().second;
            q.remove();

            for (int k = 0; k < 4; k++) {
                int newRow = row + delRow[k];
                int newCol = col + delCol[k];

                // Check valid, unvisited land cell
                if (newRow >= 0 && newRow < grid.length && newCol >= 0 && newCol < grid[0].length && grid[newRow][newCol] == '1' && !visited[newRow][newCol]) {
                    q.add(new Pair(newRow, newCol));
                    visited[newRow][newCol] = true;
                }
            }
        }
    }

    public int numIslands(char[][] grid) {
        int n = grid.length;
        int m = grid[0].length;

        boolean visited[][] = new boolean[n][m];

        AtomicInteger count = new AtomicInteger(0);

        IntStream
            .range(0, n)
            .forEach(
                i ->
                    IntStream
                        .range(0, m)
                        .forEach(
                            j -> {
                                if (!visited[i][j] && grid[i][j] == '1') {
                                    count.incrementAndGet();
                                    bfs(i, j, visited, grid);
                                }
                            }
                        )
            );
        return count.get();
    }
}
